### PR TITLE
Decomp func_8013A854 from code_d0a70_len_4fe0 and add UnkStructFromfunc8013A854

### PR DESCRIPTION
**Pull request by sonicspiral**
A decomp of `func_8013A854` from `code_d0a70_len_4fe0`. Also includes a new unknown struct `UnkStructFromfunc8013A854` at address `80156954`.
 
I used Ghidra and feedback from stuckpixel to infer the possible member data and size for `UnkStructFromfunc8013A854`. I have added a [TODO](https://github.com/sonicspiral/papermario/blob/11535abf11e782ee6444a39b8223ad48c25b8ae6/include/common_structs.h#L199) for this struct as it's member data needs type and size verification. Please reference this private Gist [ghidra_func_8013AA9C.c](https://gist.github.com/sonicspiral/66e46de6c832d13040eba431555e8886) to see the possible fields I noticed in `_DAT_80156954` could contain from the Ghidra analysis that was done here.

### CODE REVIEW

**Comment by sonicspiral** [comment]
Somehow the submodules in my branch have diverged, I wonder if I should got to each submobule and ` git checkout -f master && git pull` to fix this?

image

File: include/common_structs.h [hunk 1]
```diff
@@ -196,6 +196,21 @@ typedef struct Trigger {
...
+    here, each struct of size 0x7C.
+*/
+typedef struct   {
+    /* 0x00 */ char unk_0[0x10];
+    /* 0x10 */ s16 unk_10;
+    /* 0x14 */ s32 unk_14;
+    /* 0x18 */ char unk_18[0x64];
+} UnkStruct8013A854 ; // size = 0x7C
+
+typedef UnkStruct8013A854 UnkStruct8013A854List[UNKSTRUCT8013A854_SIZE];
```
Comment (ethteck) [inline include/common_structs.h:212]: I think this stuff should stay local to the c file until we actually know its purpose in the "bigger picture"

File: include/macros.h [hunk 2]
```diff
@@ -22,6 +22,7 @@
 #define MAX_MAPFLAGS 3
 
 #define MAX_SCRIPTS 128
+#define UNKSTRUCT8013A854_SIZE 90
```
Comment (ethteck) [inline include/macros.h:25]: this too

File: include/variables.h [hunk 3]
```diff
@@ -78,6 +78,7 @@ extern s16 D_8010CD12;
 extern s32 D_8010CD20;
 extern s32 D_8010EF08;
 extern s32 D_8010F094; // player shadow index?
+extern UnkStruct8013A854List* D_80156954[UNKSTRUCT8013A854_SIZE];
```
Comment (ethteck) [inline include/variables.h:81]: this too

File: src/code_d0a70_len_4fe0.c [hunk 4]
```diff
@@ -14,7 +14,14 @@ INCLUDE_ASM(s32, "code_d0a70_len_4fe0", func_8013A6E8);
 
 INCLUDE_ASM(s32, "code_d0a70_len_4fe0", func_8013A704);
 
-INCLUDE_ASM(s32, "code_d0a70_len_4fe0", func_8013A854);
+void func_8013A854(u32 i) {
+    UnkStruct8013A854* unkStruct80156954ListPtr;
```
Comment (ethteck) [inline src/code_d0a70_len_4fe0.c:18]: Empty line after declaration please

File: undefined_syms.txt [hunk 5]
```diff
@@ -172,6 +172,7 @@ D_80151328 = 0x80151328;
 gCurrentDoorSoundsSet = 0x80151340;
 
 gTriggerCount = 0x80151334;
+D_80156954 = 0x80156954;
```
Comment (ethteck) [inline undefined_syms.txt:175]: Please also add an entry to symbol_addrs.txt. I'll be making the formats more similar soon..

**Review by ethteck — CHANGES_REQUESTED** [review]

File: include/common_structs.h (hunk 1 above)
Comment (nanaian) [reply include/common_structs.h:212]: Or in a parallel .h file

File: undefined_syms.txt (hunk 5 above)
Comment (nanaian) [reply undefined_syms.txt:175]: What if we wrote a tool to convert between the two? I've definitely forgotten to add to symbol_addrs.txt multiple times since we added it

File: src/code_d0a70_len_4fe0.c (hunk 4 above)
Comment (nanaian) [inline src/code_d0a70_len_4fe0.c:18]: Is this required to match? Can you not do this:
```c
if (i < 90) {
    (*D_80156954)[i].unk_14 = 0;
    // ...
}
```
That dereference of D_80156954 will probably become a macro similar to GAME_STATUS once we work out what the struct is.

File: include/common_structs.h (hunk 1 above)
Comment (sonicspiral) [reply include/common_structs.h:212]: Got it. Place all of this in the c file right?

```
typedef struct {
    /* 0x00 */ char unk_0[0x10];
    /* 0x10 */ s16 unk_10;
    /* 0x14 */ s32 unk_14;
    /* 0x18 */ char unk_18[0x64];
} UnkStruct8013A854 ; // size = 0x7C

typedef UnkStruct8013A854 UnkStruct8013A854List[UNKSTRUCT8013A854_SIZE];
```

File: src/code_d0a70_len_4fe0.c (hunk 4 above)
Comment (sonicspiral) [reply src/code_d0a70_len_4fe0.c:18]: After the `void func_8013A854(u32 i) {` declaration right?

```

void func_8013A854(u32 i) {
    // <------- here
    UnkStruct8013A854* unkStruct80156954ListPtr;
    if (i < 90) {
        unkStruct80156954ListPtr = *D_80156954;
        unkStruct80156954ListPtr[i].unk_14 = 0;
        unkStruct80156954ListPtr[i].unk_10 = -1;
    }
}

```

File: undefined_syms.txt (hunk 5 above)
Comment (sonicspiral) [reply undefined_syms.txt:175]: Only add `D_80156954;0x80156954` there or add `func_8013A854;0x8013A854` too?

File: src/code_d0a70_len_4fe0.c (hunk 4 above)
Comment (sonicspiral) [reply src/code_d0a70_len_4fe0.c:18]: That's a good syntax optimization, I'd like to try and implement it, currently I get this error with these changes.

```
> ./diff.py -mw func_8013A854
cpp -Iinclude -D _LANGUAGE_C -ffreestanding -DF3DEX_GBI_2 src/code_d0a70_len_4fe0.c | tools/cc1 -O2 -quiet -G 0 -mcpu=vr4300 -mfix4300 -mips3 -mgp32 -mfp32 -o - | tools/mips-nintendo-nu64-as -EB -Iinclude -G 0 - -o build/src/code_d0a70_len_4fe0.o
src/code_d0a70_len_4fe0.c: In function `func_8013A854':
src/code_d0a70_len_4fe0.c:37: request for member `unk_14' in something not a structure or union
src/code_d0a70_len_4fe0.c:38: request for member `unk_10' in something not a structure or union
```
image

```c
typedef struct {
    /* 0x00 */ char unk_0[0x10];
    /* 0x10 */ s16 unk_10;
    /* 0x14 */ s32 unk_14;
    /* 0x18 */ char unk_18[0x64];
} UnkStruct8013A854 ; // size = 0x7C

typedef UnkStruct8013A854 UnkStruct8013A854List[UNKSTRUCT8013A854_SIZE];

extern UnkStruct8013A854List* D_80156954[UNKSTRUCT8013A854_SIZE];

void func_8013A854(u32 i) {    
    if (i < 90) {
        (*D_80156954)[i].unk_14 = 0;
        (*D_80156954)[i].unk_10 = -1;
    }
}
```

Comment (sonicspiral) [reply src/code_d0a70_len_4fe0.c:35]: @nanaian could I resolve this comment without the optimization or should I try to get this to work?

Comment (nanaian) [reply src/code_d0a70_len_4fe0.c:35]: Go ahead and resolve it. Although, I thought `D_80156954` was a pointer to an array, so this _should_ work...

Comment (sonicspiral) [reply src/code_d0a70_len_4fe0.c:35]: Okay no worries, yeah I though so too. I'm not sure if my syntax is off or not, I like your suggestion.

**Comment by sonicspiral** [comment]
Thanks for the feedback everyone. I'll leave this open for now, not sure why it still shows changes requested.
Going to work on `code_18F340` `activate_defend_command` next.

File: src/code_d0a70_len_4fe0.c [hunk 6]
```diff
@@ -14,7 +16,29 @@ INCLUDE_ASM(s32, "code_d0a70_len_4fe0", func_8013A6E8);
 
 INCLUDE_ASM(s32, "code_d0a70_len_4fe0", func_8013A704);
 
-INCLUDE_ASM(s32, "code_d0a70_len_4fe0", func_8013A854);
+/* TODO sonicspiral oct 2020
```
Comment (pixel-stuck) [inline src/code_d0a70_len_4fe0.c:19]: This comment seems irrelevant now

**Review by pixel-stuck — CHANGES_REQUESTED** [review]

File: src/code_d0a70_len_4fe0.c (hunk 6 above)
Comment (sonicspiral) [reply src/code_d0a70_len_4fe0.c:19]: Sounds good I will remove this.
